src/boot/fwupd.c

```diff
@@ -0,0 +1,576 @@
+                *updates = xrealloc(*updates,
+                                    *n_updates * sizeof(FwupUpdateTable),
+                                    (*n_updates + 1) * sizeof(FwupUpdateTable));
+                (*updates)[(*n_updates)++] = TAKE_PTR(update);
+        }
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS get_graphics_protocol_mode(EFI_HANDLE image, uint32_t *ret_mode) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles = 0;
+        static uint32_t mode = 0;
+        static int found = -1;
+        EFI_STATUS err;
+
+        assert(ret_mode);
+
+        /* Doesn't depend on the capsule being inspected, cache the result */
+        if (found > 0) {
+                *ret_mode = mode;
+                return EFI_SUCCESS;
+        }
+        if (found == 0)
+                return EFI_UNSUPPORTED;
+
+        err = BS->LocateHandleBuffer(
+                ByProtocol,
+                MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                /* SearchKey= */ NULL,
+                &n_handles,
+                &handles);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Failed to locate graphics output protocol handles: %m");
+        if (n_handles == 0 || !handles) {
+                log_error("No graphics output protocol handles found for UX capsule");
+                return EFI_UNSUPPORTED;
+        }
+
+        FOREACH_ARRAY(h, handles, n_handles) {
+                EFI_GRAPHICS_OUTPUT_PROTOCOL *proto;
+
+                err = BS->OpenProtocol(*h,
+                                       MAKE_GUID_PTR(EFI_GRAPHICS_OUTPUT_PROTOCOL),
+                                       (void **)&proto,
+                                       image,
+                                       /* ControllerHandler= */ NULL,
+                                       EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get graphics output protocol: %m");
+
+                *ret_mode = mode = proto->mode->mode;
+                found = 1;
+                return EFI_SUCCESS;
+        }
+
+        found = 0;
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_DEVICE_PATH *device_path_instance(EFI_DEVICE_PATH **device_path, size_t *ret_size) {
+        EFI_DEVICE_PATH *head, *next, *dp;
+
+        assert(device_path);
+        assert(ret_size);
+
+        if (!*device_path) {
+                *ret_size = 0;
+                return NULL;
+        }
+
+        dp = head = *device_path;
+
+        for (;;) {
+                next = device_path_next_node(dp);
+
+                if (device_path_is_end(dp))
+                        break;
+
+                dp = next;
+        }
+
+        if (dp->SubType == END_ENTIRE_DEVICE_PATH_SUBTYPE)
+                next = NULL;
+
+        *device_path = next;
+        *ret_size = ((uint8_t *) dp) - ((uint8_t *) head);
+
+        return head;
```

Comment:
can you explain what this function does, i don#t get it?

Reply:
iterates through all the subtypes, so that we can find the one we want